### DISCUSSION TITLE
Howto assign a value to a bunch of HASPmota objects

### DISCUSSION THREAD

**Discussion by waltkind** [arendst/Tasmota#20265] [Q&A] [answered]
I know how to assign a value to an HASPobject by berry-script. But if I want to assign for example 20 values to the objects p1b1 to p1b20.
`for i:10 .. 20
p1bi.x = 5 * i
end` does not work.

**Comment by sfromis** [comment, marked as answer]
`global.member(f'p1b{i}').x = 5 * i`

**Reply by waltkind** [reply]
Thank you. That's it.

**Reply by s-hadinger** [reply]
Nice!

**Reply by s-hadinger** [reply]
Alternatively you can do the following (although less readable):

`global.(f'p1b{i}').x = 5 * i`

**Reply by waltkind** [reply]
After reading the berry.wiki I was convinced that I have to work with 'Virtual members' but I was not able to read the right notation.
There are still too little working samples for HASPmota in the net, so I have to bother you with my questions.

**Reply by s-hadinger** [reply]
Agreed. We need to grow the variety of examples. Thanks for your contribution

**Comment by waltkind** [comment]
Since the command above works perfect I need your knowledge again for the opposite direction.
I have ten buttons that publish values to the mqtt-server. 
For that I have ten oneliners as:`tasmota.add_rule("hasp#p2b5#val", / value -> mqtt.publish('tele/lanbon_1/Sh_S_St', str(value)))`.
Since the amount of buttons is growing I am looking for a command to solve this with variables and a mapping list.
I tried it with different wilcard solutions like `"hasp#p2b.*#val"` but no luck. What is the right notation?

**Reply by s-hadinger** [reply]
You can't filter at pattern level. You could do `hasp#?#val` to trigger on any Haspmota event that has a `val` and test in the berry code if the pattern is in the form `hasp#p2b.*#val` for example with a regex `hasp#p2b[0-9]+#val`

**Reply by waltkind** [reply]
`tasmota.add_rule("hasp#?#val", / value -> print(value))` shows `true` and `false`. 
That is the value, but what variable represents the name of the button?

**Reply by barbudor** [reply]
Read again the doc, the function can have up to 3 argument
- value
- variable which made the trigger
- complete map of the json of the event

   https://tasmota.github.io/docs/Berry/#rules   


```
tasmota.add_rule("hasp#?#val", / value, trigger-> print(value, trigger))
```

**Reply by waltkind** [reply]
Thanks, but that doesn't show me much. On testing your command I get:
`true hasp#?#val` 
I have to add the msg. 
So `tasmota.add_rule("hasp#?#val", / value, trigger, msg -> print(value, trigger, msg))` 
shows `true hasp#?#val {'hasp': {'p2b15': {'val': true, 'event': 'changed'}}}`.

Now I can try to excerpt the 'p2b15'.  I'll do my best.

**Reply by sfromis** [reply]
You got `true` because the value of `val` in the matched object is `true`. 

`msg['hasp'].keys()()` is one trick to get the event name of `p2b15`
It gets an iterator function, and then calls it once for the first value. Not great if there is not exactly one value.

**Reply by waltkind** [reply]
Works perfect. Thank you

**Comment by waltkind** [comment]
Thanks to your help I could reduce the lines to two lines:

> m2 = {'p2b5':'Sh_S_St', 'p2b10':'Sh_S_Fr', 'p2b15':'Sh_W_In',
     'p2b20':'Sh_S_Wz', 'p2b25':'Sh_W_Ez', 'p3b5':'Sh_S_KK', 'p3b10':'Sh_S_WK'}

> tasmota.add_rule("hasp#?#val", / value, trigger, msg -> mqtt.publish('tele/lanbon_1/'+m2[msg['hasp'].keys()()], str(value)))

This will help to keep the code smaller, even when the number of buttons will increase.

For displaying the results I need the reverse mapping.
> m1 = {"Sh_S_St":'p2b5', "Sh_S_Fr":'p2b10', "Sh_W_In":'p2b15', 
      "Sh_S_Wz":'p2b20', "Sh_W_Ez":'p2b25', 'Sh_S_KK':'p3b5', 'Sh_S_WK':'p3b10'}
Is there a possibility to do this automatic or a command to use the map reverse?

**Reply by sfromis** [reply]
Not sure how much you mean by "reverse mapping", but if you only mean a map getting from a value like `Sh_S_St` to `p2b5`, this could work by putting the object into the map, instead of a string value which is the name of the object, `'Sh_S_St':p2b5` instead of `'Sh_S_St':'p2b5'`, and then looking up the name gets you directly to the global var of the object. The key here is that putting an object into a map value is not a copy, but a reference to the original object.

Of course, if you have `p2b5` as a string in `name`, then `global.member(name)` again gets you to the object.

You could also avoid having two maps, doing something like:
```
m2 = {'p2b5':'Sh_S_St', 'p2b10':'Sh_S_Fr', 'p2b15':'Sh_W_In',
'p2b20':'Sh_S_Wz', 'p2b25':'Sh_W_Ez', 'p3b5':'Sh_S_KK', 'p3b10':'Sh_S_WK'}
m1 = {} for key:m2.keys() m1[m2[key]]=key end
```
Of course, this could also be combined with putting the object into the map instead of the name.
`m1 = {} for key:m2.keys() m1[m2[key]]=global.member(key) end`

**Reply by sfromis** [reply]
Not sure how general the following is, but one way of having a naming in the `.jsonl` file would be to have your own name in `action`, like having your `p2b5` include `"action":"Sh_S_St"` among the other values.

Then `hobj = global.member(msg['hasp'].keys()())` and `hobj.action` gets you the name you put into the HASPmota object. This also aids your readability of the `.jsonl` file, and means that you basically need no name mapping in Berry.

**Reply by waltkind** [reply]
Now the line:
 `tasmota.add_rule("hasp#?#val", / value, trigger, msg -> mqtt.publish('tele/lanbon_1/'+(global.member(msg['hasp'].keys()())).action, str(value)))`
 also works and I need no map.

**Reply by sfromis** [reply]
Depends on whether you have a need for the reverse mapping you were asking about. To some extent, it could be doable to build such a reverse map by looping through the global objects determined to be HASPmota, and put them into a map indexed by `action`.

**Reply by waltkind** [reply]
I think with the solution above there is no more need for a reverse Map.
